src/markdown/generator.ml

```diff
@@ -0,0 +1,327 @@
+open Odoc_document
+open Types
+open Doctree
+
+module Markup = struct
+  type t =
+    | Inline of string list
+    | Block of t list
+    | Concat of t list
+    | Break
+    | Anchor of string
+    | String of string
+    | Env of indent * string * t * string
+    | Command of string * t
+    | OpenSqBracket
+    | CloseSqBracket
+    | OpenParenthesis
+    | CloseParenthesis
+
+  and indent = Any
+
+  let env indent s s' t = Env (indent, s, t, s')
+
+  let noop = Concat []
+
+  let break = Break
+
+  let open_sq_bracket, close_sq_bracket = (OpenSqBracket, CloseSqBracket)
+
+  let open_parenthesis, close_parenthesis = (OpenParenthesis, CloseParenthesis)
+
+  let append t1 t2 =
+    match (t1, t2) with
+    | Concat l1, Concat l2 -> Concat (l1 @ l2)
+    | Concat l1, e2 -> Concat (l1 @ [ e2 ])
+    | e1, Concat l2 -> Concat (e1 :: l2)
+    | e1, e2 -> Concat [ e1; e2 ]
+
+  let ( ++ ) = append
+
+  let concat = List.fold_left ( ++ ) (Concat [])
+
+  let rec intersperse ~sep = function
+    | [] -> []
+    | [ h ] -> [ h ]
+    | h1 :: (_ :: _ as t) -> h1 :: sep :: intersperse ~sep t
+
+  let list ?(sep = Concat []) l = concat @@ intersperse ~sep l
+
+  let str fmt = Format.ksprintf (fun s -> String s) fmt
+
+  let escaped fmt = Format.ksprintf (fun s -> String s) fmt
+
+  let command s t = Command (s, t)
+
+  let rec pp fmt t =
+    match t with
+    | Inline s -> Format.fprintf fmt "%s" (String.concat "" s)
+    | Block b ->
+        let inner = function
+          | [] -> ()
+          | [ x ] -> Format.fprintf fmt "%a" pp x
+          | x :: xs -> Format.fprintf fmt "%a@\n%a" pp x pp (Block xs)
+        in
+        inner b
+    | Concat l -> pp_many fmt l
+    | Break -> Format.fprintf fmt "@\n"
+    | Anchor s -> Format.fprintf fmt "<a id=\"%s\"></a>" s
+    | String s -> Format.fprintf fmt "%s" s
+    | Env (indent, s, t, s') ->
```

Comment:
Perhaps there should be newlines before and after `t` in case it's a `Block` ? The `Concat` constructor makes it harder.

Thought about the `Concat` constructor: Remove it, use `Block` and `Inline` instead. Change `Inline` into `Inline of t list`. Change `++` to be smart and use `Block` if a direct children is a `Block _` or `Env (_, _, Block _, _)`, use `Inline` if it's anything else.

Reply:
I don’t clearly understand why we need to perform this change!

Reply:
I think I misunderstood how inline and block works in your code.

Can you explain what the `Env` constructor does and what is expected when `t` is a `Block _` or a `Inline _` ?

Reply:
I was being used for formatting text descriptions, for example setting `some text` in source code style by wrapping it in ` [ ]`. However, the `Env` constructor and its operation are no longer being used.